src/niri.rs

```diff
@@ -3728,47 +3736,162 @@ impl Niri {
+
+        let render_result = Self::render_for_screencopy_internal(
+            renderer,
+            output,
+            &elements,
+            screencopy.with_damage(),
+            damage_tracker,
+            &screencopy,
+        )
+        .map(|_| ());
+
+        match render_result {
+            Ok(()) => screencopy.submit(false),
+            Err(_) => {
+                // recreate damage tracker to report full damage next check.
+                *damage_tracker = OutputDamageTracker::new((0, 0), 1.0, Transform::Normal);
+            }
+        }
+        render_result
+    }
+
+    fn render_for_screencopy_internal<'a>(
+        renderer: &mut GlesRenderer,
+        output: &Output,
+        elements: &[OutputRenderElements<GlesRenderer>],
+        with_damage: bool,
+        damage_tracker: &'a mut OutputDamageTracker,
+        screencopy: &Screencopy,
+    ) -> anyhow::Result<Option<&'a Vec<Rectangle<i32, Physical>>>> {
+        let OutputModeSource::Static {
+            size: last_size,
+            scale: last_scale,
+            transform: last_transform,
+        } = damage_tracker.mode().clone()
+        else {
+            unreachable!("output always has static mode");
+        };
 
-                screencopy.submit(false);
+        let size = screencopy.buffer_size();
+        let scale: Scale<f64> = output.current_scale().fractional_scale().into();
+        let transform = output.current_transform();
+
+        if size != last_size || scale != last_scale || transform != last_transform {
+            *damage_tracker = OutputDamageTracker::new(size, scale, transform);
+        }
 
-                Ok(())
+        let region_loc = screencopy.region_loc();
+        let elements = elements
+            .iter()
+            .map(|element| {
+                RelocateRenderElement::from_element(
+                    element,
+                    region_loc.upscale(-1),
+                    Relocate::Relative,
+                )
             })
-            .context("primary renderer is missing")?
+            .collect::<Vec<_>>();
+
+        // Just checked damage tracker has static mode
+        let damages = damage_tracker.damage_output(1, &elements).unwrap().0;
+        if with_damage && damages.is_none() {
+            return Ok(None);
+        }
+
+        let elements = elements.iter().rev();
+
+        match screencopy.buffer() {
+            ScreencopyBuffer::Dmabuf(dmabuf) => {
+                render_to_dmabuf(renderer, dmabuf.clone(), size, scale, transform, elements)
+                    .context("error rendering to screencopy dmabuf")
+                    .map(|_| ())
+            }
+            ScreencopyBuffer::Shm(wl_buffer) => {
+                render_to_shm(renderer, wl_buffer, scale, transform, elements)
+                    .context("error rendering to screencopy shm buffer")
+            }
+        }?;
+        renderer.unbind().context("error unbinding screencopy")?;
```

Comment:
Manual unbind shouldn't be necessary?

Reply:
I will double check this but it did cause the ghosting issue we talked about before.

Reply:
Without manual unbind there is still the ghosting like artifacts. I suspect there might be a bug in the existing code where it is not being rebound properly.